test/apps/iodemo/io_demo.cc

```diff
@@ -246,46 +379,207 @@ class P2pDemoCommon : public UcxContext {
+
+        void set_buffer(void *buffer)
+        {
+            _buffer = buffer;
         }
 
-        inline void *buffer(size_t offset = 0) const {
+        inline void *buffer(size_t offset = 0) const
+        {
             return (uint8_t*)_buffer + offset;
         }
 
-        inline void resize(size_t size) {
+        inline void resize(size_t size)
+        {
             assert(size <= _capacity);
             _size = size;
         }
 
-        inline size_t size() const {
+        inline size_t capacity() const
+        {
+            return _capacity;
+        }
+
+        inline size_t size() const
+        {
             return _size;
         }
 
+    protected:
+        const size_t _capacity;
+
+    private:
+        void *_buffer;
+        size_t _size;
+    };
+
+    class Buffer {
     public:
-        const size_t         _capacity;
+        Buffer(size_t buffer_size, BufferMemoryPool<Buffer> &pool,
+               ucs_memory_type_t memory_type) :
+            _pool(pool)
+        {
+            switch (memory_type) {
+            case UCS_MEMORY_TYPE_HOST:
+                _impl = new HostBufferImpl(buffer_size);
+                break;
+#ifdef HAVE_CUDA
+            case UCS_MEMORY_TYPE_CUDA:
+                _impl = new CudaBufferImpl(buffer_size);
+                break;
+            case UCS_MEMORY_TYPE_CUDA_MANAGED:
+                _impl = new CudaManagedBufferImpl(buffer_size);
+                break;
+#endif
+            default:
+                LOG << "ERROR: Unsupported memory type requested: "
+                    << memory_type_to_string(memory_type);
+                abort();
+            }
+
+            _impl->init();
+        }
+
+        void *buffer()
+        {
+            return _impl->buffer();
+        }
+
+        size_t capacity() const
+        {
+            return _impl->capacity();
+        }
+
+        size_t size() const
+        {
+            return _impl->size();
+        }
+
+        inline void resize(size_t size)
+        {
+            _impl->resize(size);
+        }
+
+        void release()
+        {
+            _pool.put(this);
+        }
+
+
+        ~Buffer()
+        {
+            delete _impl;
+        }
 
     private:
-        void*                     _buffer;
-        size_t                    _size;
-        MemoryPool<Buffer, true>& _pool;
+        BufferMemoryPool<Buffer> &_pool;
+        BufferImpl *_impl;
     };
 
+
+    class HostBufferImpl : public BufferImpl {
+    public:
+        HostBufferImpl(size_t size) : BufferImpl(size)
+        {
+        }
+
+        virtual void init()
+        {
+            void *tmp = memalign(ALIGNMENT, _capacity);
+
+            if (tmp == NULL) {
+                throw std::bad_alloc();
+            }
+
+            set_buffer(tmp);
+        }
+
+        ~HostBufferImpl()
+        {
+            free(buffer());
+        }
+    };
+
+#ifdef HAVE_CUDA
+    class CudaMemoryBufferImpl : public BufferImpl {
```

Comment:
do we really need the Buffer/BufferImpl/... hierarchy?
can it be done simpler?
we need to remember this is also demo for users so it should be as short as possible

Reply:
I can suggest we have a single class, which would hold the memory type as a member.
Would that be better, what to do you think? @yosefe  @evgeny-leksikov

Reply:
IMHO all-in-one-super-duper-class does not make things simpler. Probably we can separate current "Buffer" class to static fabric function and rename "BufferImpl" to base "Buffer" class to track in Pools. I would keep separate implementations in separate classes.